### PR TITLE
Added asVector3(), asPosition() and asLocation()

### PR DESCRIPTION
**Pull request by matcracker**
## Introduction

This pull request allow developers to get faster access to the block position without work around the level or its coordinates.
### API changes
Added `Position::asPosition()` function.
Added `Location::asLocation()` function.
Added `Vector3::asVector3()` function.
Deprecated `Entity::getLocation()` function. _(Note by @SOF3: Undone while squash-merging)_
Deprecated `Entity::getPosition()` function. _(Note by @SOF3: Undone while squash-merging)_

### CODE REVIEW

**Comment by Muqsit** [comment]
class Block extends Position
Isn't that enough?

**Comment by dktapps** [comment]
Why? A Block _is_ a position.

You also cannot assume that any of these fields are defined when this method is called.

Besides, if you _really_ want a position, just use `Position::fromObject($block, $block->getLevel())`.

**Comment by matcracker** [comment]
Uh huh, I didn't know there was a `Position::fromObject($block, $block->getLevel())`, good to know. Anyway, in my option, it's an hidden method. See my next commit, would be better.

**Comment by dktapps** [comment]
> You also cannot assume that any of these fields are defined when this method is called.

you still didn't cater for this, and I object to adding extra API methods unless they actually contribute something useful to the API.

**Comment by SOF3** [comment]
I recommend adding the `getLocation()`, `getPosition()` and `getVector3()` methods to those three classes directly, since similar methods also exist in Player class etc.

**Comment by matcracker** [comment]
According to `getLocation()` constructor, it requires yaw and pitch that those blocks haven't got.
`Position` extends `Vector3`, so it's useless to get the `gerVector3()` method.

**Comment by Muqsit** [comment]
@SOF3 has a better idea.
Add getPosition() in Location,
Add getVector3() in Position.
Done.

**Comment by matcracker** [comment]
@Muqsit and where do you add `getLocation()`?
Summary:
`Position::getVector3()`
`Position::getLocation()`
`Location::getPosition()`

**Comment by Muqsit** [comment]
@matcracker you dont need to add it.

**Comment by IrishPacks** [comment]
A couple of months ago I was working on a plugin that dealt with storing tons of data in an array(Up to 12000 blocks at least). I was just storing the block, and I noticed how bad the performance was, so what I did was instead of storing the block, I decided to create a new Vector3 and store that instead, which lead to MASSIVE performance increases.

Long story short, Block::getPosition() should be made a function for situations like those. As stated before, blocks ARE positions, but they carry massive amounts of unneeded data when storing in bulk.

**Comment by dktapps** [comment]
> Besides, if you really want a position, just use Position::fromObject($block, $block->getLevel()).

I don't see the point in adding another method for this.

**Comment by Muqsit** [comment]
@IrishPacks Why only Block::getPosition()? Why not Entity::getPosition() and every single thing that'll extend Position in the future. I don't know why SOF3's idea doesn't make sense, it's much more flexible!

**Comment by IrishPacks** [comment]
@Muqsit I did skim those, but that would make more sense to apply it to anything that is a position, vector3, etc. :+1: from me

**Comment by SOF3** [comment]
Something that extends Position doesn't necessarily extend Location. It is more useful to create these:

```php
public function Vector3::getVector3() : Vector3;
public function Position::getPosition() : Position;
public function Location::getLocation() : Location;
```

However, it may be more descriptive to call them `asVector3()`, `asPosition()` and `asLocation()` since the function actually returns an equal value, not a value that it owns.

**Comment by matcracker** [comment]
@SOF3 What do they return? A self instance or a new object instance?

**Comment by SOF3** [comment]
@matcracker you're referring to a this instance, not self instance. 
And my answer is, it should be a new instance. A getter sounds like it's fetching an existing value, while it's actually something similar to type conversion, which returns ne

File: src/pocketmine/level/Location.php [hunk 1]
```diff
@@ -57,6 +57,15 @@ public static function fromObject(Vector3 $pos, Level $level = null, $yaw = 0.0,
 		return new Location($pos->x, $pos->y, $pos->z, $yaw, $pitch, $level ?? (($pos instanceof Position) ? $pos->level : null));
 	}
 
+	/**
+	 * Return a Location instance
+	 * 
+	 * @return Location
+	 */
+	public function asLocation() : Location {
+		return new Location($this->x, $this->y, $this->z, $this->yaw, $this->pitch, $this->level);
```
Comment (Muqsit) [inline src/pocketmine/level/Location.php]: return self;

Comment (matcracker) [inline src/pocketmine/level/Location.php]: @SOF3 said:
> ... it should be a new instance. A getter sounds like it's fetching an existing value, while it's actually something similar to type conversion, which returns new objects.

Comment (dktapps) [inline src/pocketmine/level/Location.php]: That would return descendents of the class, which defeats the object of doing this. (No pun intended).

File: src/pocketmine/level/Location.php [hunk 2]
```diff
@@ -57,6 +57,15 @@ public static function fromObject(Vector3 $pos, Level $level = null, $yaw = 0.0,
 		return new Location($pos->x, $pos->y, $pos->z, $yaw, $pitch, $level ?? (($pos instanceof Position) ? $pos->level : null));
 	}
 
+	/**
+	 * Return a Location instance
+	 * 
+	 * @return Location
+	 */
+	public function asLocation() : Location {
```
Comment (Muqsit) [inline src/pocketmine/level/Location.php]: code formatting

File: src/pocketmine/level/Position.php [hunk 3]
```diff
@@ -47,6 +47,15 @@ public static function fromObject(Vector3 $pos, Level $level = null){
 	}
 
 	/**
+	 * Return a Position instance
+	 * 
+	 * @return Position
+	 */
+	public function asPosition() : Position {
```
Comment (Muqsit) [inline src/pocketmine/level/Position.php]: code formatting

File: src/pocketmine/math/Vector3.php [hunk 4]
```diff
@@ -158,6 +158,15 @@ public function getSide($side, $step = 1){
 	}
 
 	/**
+	 * Return a Vector3 instance
+	 * 
+	 * @return Vector3
+	 */
+	public function asVector3() : Vector3 {
```
Comment (Muqsit) [inline src/pocketmine/math/Vector3.php]: code formatting

Comment (Muqsit) [inline src/pocketmine/math/Vector3.php]: code formatting

**Comment by Muqsit** [comment]
Um, what about [these](https://github.com/pmmp/PocketMine-MP/blob/master/src/pocketmine/entity/Entity.php#L1386-#L1392)? I don't think we'll be needing them anymore. But if we remove it, we'll need to change a lot of `Entity::getPosition()` and `Entity::getLocation()` functions to the new `as*()` functions.... unless we name the new functions `get*()`.

**Comment by matcracker** [comment]
This is why I initially wanted to implement just getPosition() on Block class.

**Comment by SOF3** [comment]
@Muqsit \x42lameShighi, the answer to life, the universe and everything.
Deprecating the Entity getLocation() methods might be reasonable, but there is no strong reason to remove them soon.

File: src/pocketmine/math/Vector3.php [hunk 5]
```diff
@@ -158,6 +158,15 @@ public function getSide($side, $step = 1){
...
+		return new Vector3($this->x, $this->y, $this->z);
```
Comment (thebigsmileXD) [inline src/pocketmine/math/Vector3.php]: Doesn't this return itself? why not just clone self;?

Comment (dktapps) [inline src/pocketmine/math/Vector3.php]: as I said above...

> That would return descendents of the class, which defeats the object of doing this. (No pun intended).

Comment (thebigsmileXD) [inline src/pocketmine/math/Vector3.php]: But xxxclass as xxxclass is just useless.

Comment (SOF3) [inline src/pocketmine/math/Vector3.php]: @thebigsmileXD this whole pull request adds functions that serve the exact same purpose as the `Player::getPosition()` method.

File: src/pocketmine/entity/Entity.php [hunk 6]
```diff
@@ -1383,10 +1383,20 @@ protected function switchLevel(Level $targetLevel){
...
+	 * @deprecated Use asPosition() instead getPosition()
+	 * 
+	 * @return Position
+	 */
 	public function getPosition(){
 		return new Position($this->x, $this->y, $this->z, $this->level);
 	}
 
+	/**
+	 * @deprecated Use asLocation() instead getLocation()
```
Comment (Muqsit) [inline src/pocketmine/entity/Entity.php]: That being said, you'll need to change all deprecated Entity::get*() fields to Entity::as*() fields... pain! :/

Comment (SOF3) [inline src/pocketmine/entity/Entity.php]: @Muqsit there are actually many places in the PocketMine source that should use this method but did not do so.

Comment (dktapps) [inline src/pocketmine/entity/Entity.php]: I don't think there's any need for deprecating these. While entities are indeed positions, it also does make sense to be able to say Entity->getPosition() or Entity->getLocation(). The new methods are intended to make life easier for developers, yes?

**Comment by dktapps** [comment]
This has been merged in the referenced commit with some minor alterations. Thanks for contributing.

**Pull request closed without merging**